### PR TITLE
Volume persistence incremental

### PR DESCRIPTION
**Pull request by heavenlyhash**
Several more isolated changes that pave the way for volume persistence.

Additional testing coming along with #924, which makes use of these changes.

### CODE REVIEW

File: host/host.go [hunk 1]
```diff
@@ -161,19 +162,21 @@ func runDaemon(args *docopt.Args) {
...
-	volProv, err := zfsVolume.NewProvider(&zfsVolume.ProviderConfig{
-		DatasetName: "flynn-default",
-		Make: &zfsVolume.MakeDev{
-			BackingFilename: "/var/lib/flynn/volumes/zfs/vdev/flynn-default-zpool.vdev",
-			Size:            int64(math.Pow(2, float64(30))),
-		},
-	})
-	if err != nil {
-		log.Fatalf("unable to create volume backend: %s", err)
+	volProvFn := func() (volume.Provider, error) {
```
Comment (titanous) [inline host/host.go]: Any reason to name this function over passing it inline?

File: host/volume/manager/manager.go [hunk 2]
```diff
@@ -30,20 +30,30 @@ type Manager struct {
...
-	return &Manager{
-		defaultProvider: p,
-		providers:       map[string]volume.Provider{"default": p},
-		volumes:         map[string]volume.Volume{},
-		namedVolumes:    map[string]string{},
+func New(defProvFn func() (volume.Provider, error)) (*Manager, error) {
+	m := &Manager{
+		providers:    map[string]volume.Provider{},
+		volumes:      map[string]volume.Volume{},
+		namedVolumes: map[string]string{},
```
Comment (titanous) [inline host/volume/manager/manager.go]: Use `make` for each of these.

File: host/volume/manager/manager.go [hunk 3]
```diff
@@ -30,20 +30,30 @@ type Manager struct {
...
+	}
+	if _, ok := m.providers["default"]; !ok {
+		p, err := defProvFn()
```
Comment (titanous) [inline host/volume/manager/manager.go]: This path looks like it is hit 100% of the time?

File: host/volume/manager/manager.go [hunk 4]
```diff
@@ -30,20 +30,30 @@ type Manager struct {
 	namedVolumes map[string]string
 }
 
-func New(p volume.Provider) *Manager {
-	return &Manager{
-		defaultProvider: p,
-		providers:       map[string]volume.Provider{"default": p},
-		volumes:         map[string]volume.Volume{},
-		namedVolumes:    map[string]string{},
+func New(defProvFn func() (volume.Provider, error)) (*Manager, error) {
```
Comment (titanous) [inline host/volume/manager/manager.go]: Why pass in a function?

File: host/volume/manager/manager.go (hunk 3 above)
Comment (heavenlyhash) [inline host/volume/manager/manager.go]: Correct; this is one of the parts that makes more sense with the next step in #924.  See https://github.com/flynn/flynn/pull/924/files#diff-dbf2e862dcaac249b6826c1bb5263118R91

Comment (titanous) [inline host/volume/manager/manager.go]: Ah, got it.

**Comment by heavenlyhash** [comment]
Comments addressed.

**Comment by titanous** [comment]
LGTM

**Pull request merged**